### ISSUE TITLE
Unable to train model to a state where it doesn't loses

### ISSUE THREAD

**Issue by big-bang3** [3cky/tensorflow-rl-tictactoe#1]
I made a web app which allows player to play with bot.I am using the output of q_nn model to get next move for the bot.However in some cases, player is able to defeat bot .In such cases I set the reward for previous state to -1 and reward for state where user wins to 1 and update the model. 
It allows bot to change moves for some situations but sometimes model gives same output even though I updated it with -1 reward. Please help so that I can train the model while player playes it

**Comment by 3cky** [comment]
You've probably hit a local minimum while updating the model. In such case gradients are small and training process became very slow. There are methods to mitigate such effects, like minibatching and expirience replay, but none of them is implemented (yet) in tensorflow-rl-tictactoe.

**Comment by big-bang3** [comment]
I am actually new to machine learning.From what I have studied about experience replay, instead of updating the model instantly, I have to store states for various multiple situations where bot loses and then train the model with random set of batches from them at some interval. Is it right ?

Also, local minimum, does it means that model is unable to update in such a way that the loss minimizes in following lines? 

     loss = tf.reduce_mean(tf.square(y - action_q_values))
     optimizer = tf.train.AdamOptimizer(learning_rate)
     q_nn_update = optimizer.minimize(loss, var_list=tf.trainable_variables())

**Comment by 3cky** [comment]
Yes, past experience replay helps the model don't stuck in the local minima where gradients are small so model updating is very slow. Probability of this situation depends on the model complexity, but even simple models like tictactoe are tending to fall into local minima pits.

**Comment by big-bang3** [comment]
ok, thanks
